File: packages/b2c-tooling-sdk/src/clients/custom-apis.ts
/*
 * Copyright (c) 2025, Salesforce, Inc.
 * SPDX-License-Identifier: Apache-2
 * For full license text, see the license.txt file in the repo root or http://www.apache.org/licenses/LICENSE-2.0
 */
/**
 * Custom APIs DX API client for B2C Commerce.
 *
 * Provides a fully typed client for Custom APIs DX API operations using
 * openapi-fetch with OAuth authentication middleware. Used for retrieving
 * the status of deployed Custom API endpoints.
 *
 * @module clients/custom-apis
 */
import createClient, {type Client} from 'openapi-fetch';
import type {AuthStrategy} from '../auth/types.js';
import {OAuthStrategy} from '../auth/oauth.js';
import type {paths, components} from './custom-apis.generated.js';
import {createAuthMiddleware, createLoggingMiddleware} from './middleware.js';
import {globalMiddlewareRegistry, type MiddlewareRegistry} from './middleware-registry.js';

/**
 * Re-export generated types for external use.
 */
export type {paths, components};

/**
 * The typed Custom APIs client for Custom SCAPI endpoint management.
 *
 * ## Common Endpoints
 *
 * | Method | Path | Description |
 * |--------|------|-------------|
 * | GET | `/organizations/{organizationId}/endpoints` | List all custom API endpoints |
 * | GET | `/organizations/{organizationId}/endpoints/{apiName}` | Get specific API info |
 *
 * @example
 * ```typescript
 * import { createCustomApisClient, toOrganizationId } from '@salesforce/b2c-tooling-sdk/clients';
 * import { OAuthStrategy } from '@salesforce/b2c-tooling-sdk/auth';
 *
 * const auth = new OAuthStrategy({
 *   clientId: 'your-client-id',
 *   clientSecret: 'your-client-secret',
 * });
 *
 * const client = createCustomApisClient(
 *   { shortCode: 'kv7kzm78', tenantId: 'zzxy_prd' },
 *   auth
 * );
 *
 * // List all custom API endpoints
 * const { data, error } = await client.GET('/organizations/{organizationId}/endpoints', {
 *   params: { path: { organizationId: toOrganizationId('zzxy_prd') } }
 * });
 * ```
 *
 * @see {@link createCustomApisClient} for instantiation
 * @see {@link https://developer.salesforce.com/docs/commerce/commerce-api/references/custom-apis?meta=Summary | Custom APIs Reference}
 */
export type CustomApisClient = Client<paths>;

/**
 * Helper type to extract response data from an operation.
 */
export type CustomApisResponse<T> = T extends {content: {'application/json': infer R}} ? R : never;

/**
 * Standard Custom APIs error response structure.
 */
export type CustomApisError = components['schemas']['ErrorResponse'];

/** Default OAuth scopes required for Custom APIs (read-only) */
export const CUSTOM_APIS_DEFAULT_SCOPES = ['sfcc.custom-apis'];

/**
 * Configuration for creating a Custom APIs client.
 */
export interface CustomApisClientConfig {
  /**
   * The short code for the SCAPI instance.
   * This is typically a 4-8 character alphanumeric code.
   * @example "kv7kzm78"
   */
  shortCode: string;

  /**
   * The tenant ID (with or without f_ecom_ prefix).
   * Used to build the organizationId path parameter and tenant-specific OAuth scope.
   * @example "zzxy_prd" or "f_ecom_zzxy_prd"
   */
  tenantId: string;

  /**
   * Optional scope override. If not provided, defaults to domain scope
   * (sfcc.custom-apis) plus tenant-specific scope (SALESFORCE_COMMERCE_API:{tenant}).
   */
  scopes?: string[];

  /**
   * Middleware registry to use for this client.
   * If not specified, uses the global middleware registry.
   */
  middlewareRegistry?: MiddlewareRegistry;
}

/**
 * Creates a typed Custom APIs DX API client.
 *
 * Returns the openapi-fetch client directly, with authentication
 * handled via middleware. This gives full access to all openapi-fetch
 * features with type-safe paths, parameters, and responses.
 *
 * The client automatically handles OAuth scope requirements:
 * - Domain scope: `sfcc.custom-apis` (or custom via config.scopes)
 * - Tenant scope: `SALESFORCE_COMMERCE_API:{tenantId}`
 *
 * @param config - Custom APIs client configuration including shortCode and tenantId
 * @param auth - Authentication strategy (typically OAuth)
 * @returns Typed openapi-fetch client
 *
 * @example
 * // Create Custom APIs client - scopes are handled automatically
 * const oauthStrategy = new OAuthStrategy({
 *   clientId: 'your-client-id',
 *   clientSecret: 'your-client-secret',
 * });
 *
 * const client = createCustomApisClient(
 *   { shortCode: 'kv7kzm78', tenantId: 'zzxy_prd' },
 *   oauthStrategy
 * );
 *
 * // Get all Custom API endpoints
 * const { data, error } = await client.GET('/organizations/{organizationId}/endpoints', {
 *   params: {
 *     path: { organizationId: toOrganizationId('zzxy_prd') }
 *   }
 * });
 */
export function createCustomApisClient(config: CustomApisClientConfig, auth: AuthStrategy): CustomApisClient {
  const registry = config.middlewareRegistry ?? globalMiddlewareRegistry;

  const client = createClient<paths>({
    baseUrl: `https://${config.shortCode}.api.commercecloud.salesforce.com/dx/custom-apis/v1`,
  });

  // Build required scopes: domain scope + tenant-specific scope
  const requiredScopes = config.scopes ?? [...CUSTOM_APIS_DEFAULT_SCOPES, buildTenantScope(config.tenantId)];

  // If OAuth strategy, add required scopes; otherwise use as-is
  const scopedAuth = auth instanceof OAuthStrategy ? auth.withAdditionalScopes(requiredScopes) : auth;

  // Core middleware: auth first
  client.use(createAuthMiddleware(scopedAuth));

  // Plugin middleware from registry
  for (const middleware of registry.getMiddleware('custom-apis')) {
    client.use(middleware);
  }

  // Logging middleware last (sees complete request with all modifications)
  client.use(createLoggingMiddleware('CUSTOM-APIS'));

  return client;
}

/** Prefix required for SCAPI organizationId */
export const ORGANIZATION_ID_PREFIX = 'f_ecom_';

/** Prefix for tenant-specific SCAPI OAuth scopes */
export const SCAPI_TENANT_SCOPE_PREFIX = 'SALESFORCE_COMMERCE_API:';

/**
 * Ensures a tenant ID has the required f_ecom_ prefix for use as an SCAPI organizationId.
 * If the value already has the prefix, it's returned as-is.
 *
 * @param tenantId - The tenant ID (e.g., "zzxy_prd" or "f_ecom_zzxy_prd")
 * @returns The organization ID with the f_ecom_ prefix (e.g., "f_ecom_zzxy_prd")
 *
 * @example
 * toOrganizationId('zzxy_prd')        // Returns 'f_ecom_zzxy_prd'
 * toOrganizationId('f_ecom_zzxy_prd') // Returns 'f_ecom_zzxy_prd' (unchanged)
 */
export function toOrganizationId(tenantId: string): string {
  if (tenantId.startsWith(ORGANIZATION_ID_PREFIX)) {
    return tenantId;
  }
  return `${ORGANIZATION_ID_PREFIX}${tenantId}`;
}

/**
 * Extracts the raw tenant ID by stripping the f_ecom_ prefix if present.
 *
 * @param value - The tenant ID or organization ID (e.g., "zzxy_prd" or "f_ecom_zzxy_prd")
 * @returns The raw tenant ID without the prefix (e.g., "zzxy_prd")
 *
 * @example
 * toTenantId('f_ecom_zzxy_prd') // Returns 'zzxy_prd'
 * toTenantId('zzxy_prd')        // Returns 'zzxy_prd' (unchanged)
 */
export function toTenantId(value: string): string {
  if (value.startsWith(ORGANIZATION_ID_PREFIX)) {
    return value.slice(ORGANIZATION_ID_PREFIX.length);
  }
  return value;
}

/**
 * Builds the tenant-specific OAuth scope required for SCAPI APIs.
 *
 * @param tenantId - The tenant ID (with or without f_ecom_ prefix)
 * @returns The tenant-specific scope (e.g., "SALESFORCE_COMMERCE_API:zzxy_prd")
 *
 * @example
 * buildTenantScope('zzxy_prd')        // Returns 'SALESFORCE_COMMERCE_API:zzxy_prd'
 * buildTenantScope('f_ecom_zzxy_prd') // Returns 'SALESFORCE_COMMERCE_API:zzxy_prd'
 */
export function buildTenantScope(tenantId: string): string {
  return `${SCAPI_TENANT_SCOPE_PREFIX}${toTenantId(tenantId)}`;
}

/**
 * Constructs the full SCAPI URL for a custom API endpoint.
 *
 * This function builds the complete URL that can be used to call a custom SCAPI API endpoint.
 * The URL includes the short code, API name, version, organization ID, and endpoint path.
 *
 * @param shortCode - SCAPI shortCode (e.g., "kv7kzm78")
 * @param organizationId - Organization ID with f_ecom_ prefix (use {@link toOrganizationId} to ensure correct format)
 * @param apiName - Custom API name
 * @param apiVersion - API version (e.g., "v1")
 * @param endpointPath - Endpoint path (with or without leading slash)
 * @returns Full SCAPI URL for testing
 *
 * @example
 * ```typescript
 * import { buildCustomApiUrl, toOrganizationId } from '@salesforce/b2c-tooling-sdk/clients';
 *
 * const url = buildCustomApiUrl(
 *   "kv7kzm78",
 *   toOrganizationId("zzxy_prd"),
 *   "loyalty-info",
 *   "v1",
 *   "/customers/{customerId}/points"
 * );
 * // Returns: "https://kv7kzm78.api.commercecloud.salesforce.com/custom/loyalty-info/v1/organizations/f_ecom_zzxy_prd/customers/{customerId}/points"
 * ```
 */
export function buildCustomApiUrl(
  shortCode: string,
  organizationId: string,
  apiName: string,
  apiVersion: string,
  endpointPath: string,
): string {
  // Remove leading slash from endpoint path if present
  const cleanPath = endpointPath.startsWith('/') ? endpointPath.slice(1) : endpointPath;

  return `https://${shortCode}.api.commercecloud.salesforce.com/custom/${apiName}/${apiVersion}/organizations/${organizationId}/${cleanPath}`;
}

File: python/src/b2c_tooling_sdk/clients/custom_apis.py
# Copyright (c) 2025, Salesforce, Inc.
# SPDX-License-Identifier: Apache-2.0
# For full license text, see the LICENSE file in the repo root or http://www.apache.org/licenses/LICENSE-2.0
"""Custom APIs DX API client for B2C Commerce.

Mirrors ``src/clients/custom-apis.ts``. Provides a client for Custom APIs DX API
operations (retrieving the status of deployed Custom API endpoints) plus the
shared SCAPI tenant/organization/scope helpers (:func:`to_organization_id`,
:func:`normalize_tenant_id`, :func:`build_tenant_scope`) used across every SCAPI
domain client.

Returns a configured :class:`~b2c_tooling_sdk.clients._core.HttpClient`; callers
use ``client.get(path, {...})`` etc., each returning a
:class:`~b2c_tooling_sdk.clients._core.ClientResult`.
"""

from __future__ import annotations

from dataclasses import dataclass, field

from b2c_tooling_sdk.auth.types import AuthStrategy
from b2c_tooling_sdk.clients._core import HttpClient
from b2c_tooling_sdk.clients.middleware import create_auth_middleware, create_logging_middleware
from b2c_tooling_sdk.clients.middleware_registry import MiddlewareRegistry, global_middleware_registry
from b2c_tooling_sdk.clients.scapi_backend_utils import with_scopes

#: The typed Custom APIs client. Aliased to :class:`HttpClient` (the
#: ``openapi-fetch`` ``Client`` analog).
CustomApisClient = HttpClient

#: Default OAuth scopes required for Custom APIs (read-only).
CUSTOM_APIS_DEFAULT_SCOPES = ["sfcc.custom-apis"]

#: Prefix required for a SCAPI ``organizationId``.
ORGANIZATION_ID_PREFIX = "f_ecom_"

#: Prefix for tenant-specific SCAPI OAuth scopes.
SCAPI_TENANT_SCOPE_PREFIX = "SALESFORCE_COMMERCE_API:"


@dataclass
class CustomApisClientConfig:
    """Configuration for creating a Custom APIs client."""

    #: The short code for the SCAPI instance (typically 4-8 alphanumeric chars).
    short_code: str
    #: The tenant ID (with or without ``f_ecom_`` prefix). Used to build the
    #: ``organizationId`` path parameter and tenant-specific OAuth scope.
    tenant_id: str
    #: Optional scope override. Defaults to the domain scope plus tenant scope.
    scopes: list[str] | None = None
    #: Optional middleware registry override (mainly for tests).
    middleware_registry: MiddlewareRegistry | None = field(default=None)


def create_custom_apis_client(config: CustomApisClientConfig, auth: AuthStrategy) -> CustomApisClient:
    """Create a typed Custom APIs DX API client.

    The client automatically handles OAuth scope requirements:

    - Domain scope: ``sfcc.custom-apis`` (or custom via ``config.scopes``)
    - Tenant scope: ``SALESFORCE_COMMERCE_API:{tenant_id}``

    :param config: Custom APIs client configuration including short code and tenant ID.
    :param auth: Authentication strategy (typically OAuth).
    :returns: A configured :class:`HttpClient`.
    """
    registry = config.middleware_registry or global_middleware_registry

    client = HttpClient(
        f"https://{config.short_code}.api.commercecloud.salesforce.com/dx/custom-apis/v1",
        client_type="custom-apis",
    )

    # Build required scopes: domain scope + tenant-specific scope.
    required_scopes = config.scopes or [*CUSTOM_APIS_DEFAULT_SCOPES, build_tenant_scope(config.tenant_id)]

    # If auth supports scopes, add required scopes; otherwise use as-is.
    scoped_auth = with_scopes(auth, required_scopes)

    # Core middleware: auth first.
    client.use(create_auth_middleware(scoped_auth))

    # Plugin middleware from the registry.
    for middleware in registry.get_middleware("custom-apis"):
        client.use(middleware)

    # Logging middleware last (sees the complete request with all modifications).
    client.use(create_logging_middleware("CUSTOM-APIS"))

    return client


def to_organization_id(tenant_id: str) -> str:
    """Ensure a tenant ID has the ``f_ecom_`` prefix for use as a SCAPI ``organizationId``.

    If the value already has the prefix, it's returned as-is.

    >>> to_organization_id("zzxy_prd")
    'f_ecom_zzxy_prd'
    >>> to_organization_id("f_ecom_zzxy_prd")
    'f_ecom_zzxy_prd'
    """
    return f"{ORGANIZATION_ID_PREFIX}{normalize_tenant_id(tenant_id)}"


def normalize_tenant_id(value: str) -> str:
    """Normalize any parseable tenant/organization ID form to canonical underscore format.

    Supported input forms (all resolve to ``abcd_123``):

    - ``abcd_123`` — canonical tenant ID (returned as-is)
    - ``abcd-123`` — hyphenated tenant ID
    - ``f_ecom_abcd_123`` — organization ID
    - ``f_ecom_abcd-123`` — org ID with hyphenated tenant
    - ``abcd-123.dx.commercecloud.salesforce.com`` — sandbox hostname

    >>> normalize_tenant_id("f_ecom_zzxy_prd")
    'zzxy_prd'
    >>> normalize_tenant_id("zzxy-prd")
    'zzxy_prd'
    >>> normalize_tenant_id("zzxy-prd.dx.commercecloud.salesforce.com")
    'zzxy_prd'
    """
    tenant = value.strip()

    # Extract hostname prefix: "abcd-123.dx.commercecloud.salesforce.com" -> "abcd-123".
    if "." in tenant:
        tenant = tenant.split(".")[0]

    # Strip f_ecom_ prefix (handles org ID form).
    if tenant.startswith(ORGANIZATION_ID_PREFIX):
        tenant = tenant[len(ORGANIZATION_ID_PREFIX) :]

    # Convert hyphens to underscores.
    return tenant.replace("-", "_")


def build_tenant_scope(tenant_id: str) -> str:
    """Build the tenant-specific OAuth scope required for SCAPI APIs.

    >>> build_tenant_scope("zzxy_prd")
    'SALESFORCE_COMMERCE_API:zzxy_prd'
    >>> build_tenant_scope("f_ecom_zzxy_prd")
    'SALESFORCE_COMMERCE_API:zzxy_prd'
    """
    return f"{SCAPI_TENANT_SCOPE_PREFIX}{normalize_tenant_id(tenant_id)}"


__all__ = [
    "CUSTOM_APIS_DEFAULT_SCOPES",
    "ORGANIZATION_ID_PREFIX",
    "SCAPI_TENANT_SCOPE_PREFIX",
    "CustomApisClient",
    "CustomApisClientConfig",
    "build_tenant_scope",
    "create_custom_apis_client",
    "normalize_tenant_id",
    "to_organization_id",
]
